### PR TITLE
Add DefaultNoBound

### PR DESCRIPTION
**Pull request by kianenigma**
Will help a bit for cases like as in: https://github.com/paritytech/substrate/pull/8197/files#

### CODE REVIEW

File: frame/support/procedural/src/default_no_bound.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		},
+		syn::Data::Union(_) => {
+			let msg = "Union type not supported by `derive(CloneNoBound)`";
+			return syn::Error::new(input.span(), msg).to_compile_error().into()
+		},
+	};
+
+	quote::quote!(
+		const _: () = {
+			impl #impl_generics core::clone::Clone for #name #ty_generics #where_clause {
```
Comment (bkchr) [inline frame/support/procedural/src/default_no_bound.rs:90]: Sir? :D

**Review by bkchr — COMMENTED** [review]
And maybe some ui test?

File: frame/support/procedural/src/default_no_bound.rs (hunk 1 above)
Comment (kianenigma) [reply frame/support/procedural/src/default_no_bound.rs:90]: Copy pasta 🤦

**Comment by kianenigma** [comment]
> And maybe some ui test?

Added.

File: frame/support/procedural/src/default_no_bound.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				let variant_ident = &first_variant.ident;
+				match &first_variant.fields {
+					syn::Fields::Named(named) => {
+						let fields = named.named.iter()
+							.map(|i| &i.ident)
+							.map(|i| quote::quote_spanned!(i.span() =>
+								#i: core::default::Default::default()
+							));
+
+						quote::quote!( Self { #( #fields, )* } )
```
Comment (kianenigma) [inline frame/support/procedural/src/default_no_bound.rs:65]: errr this is still wrong.

**Review by bkchr — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by kianenigma** [comment]
bot megre

**Pull request merged**